core/src/read.jl

```diff
@@ -246,7 +246,102 @@ const conservative_nodetypes = Set{NodeType.T}([
+        db,
+        "SELECT node_id, node_type, subnetwork_id, source_priority FROM Node ORDER BY subnetwork_id, source_priority",
+    )
+
+    # Build dictionary source type -> default source priority (e.g. "user_demand" => 1000)
+    default_source_priority_dict = Dict{Symbol, Int32}()
+    source_types = propertynames(default_source_priority)
+    for source_type in source_types
+        if source_type == :boundary
+            default_priority = default_source_priority.boundary
+            default_source_priority_dict[:flow_boundary] = default_priority
+            default_source_priority_dict[:level_boundary] = default_priority
+        else
+            default_source_priority_dict[source_type] =
+                getproperty(default_source_priority, source_type)
+        end
+    end
+
+    # Get named tuples (; node_id, subnetwork_id, source_priority)
+    source_priority_tuples = SOURCE_TUPLE[]
+
+    errors = false
+
+    for row in node_rows
+        # Only source nodes that are part of a subnetwork are relevant
+        is_source = false
+        # One source priority can apply to multiple sources in the case of a LevelDemand
+        # node which connects to multiple basins
+        node_ids = [NodeID(Symbol(row.node_type), row.node_id, p)]
+        if !ismissing(row.subnetwork_id)
+            node_type = Symbol(snake_case(row.node_type))
+
+            if node_type ∈ source_types
+                # The case where the node type is also a source type
+                source_priority =
+                    coalesce(row.source_priority, default_source_priority_dict[node_type])
+                source_type = AllocationSourceType.T(node_type)
+                is_source = true
+                # If the row is for a level demand or flow demand, make node_ids
+                # for the node(s) that has/have the demand
+                if only(node_ids).type ∈ (NodeType.LevelDemand, NodeType.FlowDemand)
+                    node_ids =
+                        outneighbor_labels_type(graph, only(node_ids), LinkType.control)
+                end
+            elseif node_type ∈ (:flow_boundary, :level_boundary)
+                # The case where the node type is a boundary source node type
+                source_priority =
+                    coalesce(row.source_priority, default_source_priority_dict[node_type])
+                source_type = AllocationSourceType.boundary
+                is_source = true
+            else
+                if row.subnetwork_id != 1 # Not in the main network
+                    for id in inflow_ids(p.graph, only(node_ids))
+                        if graph[id].subnetwork_id == 1 # Connects to the main network
+                            is_source = true
+                            source_priority =
+                                default_source_priority_dict[:subnetwork_inlet]
+                            source_type = AllocationSourceType.subnetwork_inlet
+                            break
```

Comment:
This might be the deepest nested code in our codebase! 🎉 

Can we filter the table earlier (maybe even in SQL?) and fill in the defaults in a separate loop?